README.md
# .srt parser 2

## Supports null safety / Dart 3

## description
A .srt file parser that allows you to read a String containing HTML, coordination and timestamps information and convert it to a list of subtitle objects.  
Each Subtitle object contains a list of subtitle subsets with their corresponding metadata. 

### Example 

```dart
import 'package:srt_parser_2/srt_parser_2.dart';

const String data = '''1
00:02:26,407 --> 00:02:31,356  X1:100 X2:100 Y1:100 Y2:100
+ time to move on, <u><b><font color="#00ff00">Arman</font></b></u>.
- OK

2
00:02:31,567 --> 00:02:37,164 
+ Lukas is publishing his library.
- I like the man.
''';

void main() {
  List<Subtitle> subtitles = parseSrt(data);
  for (Subtitle item in subtitles) {
    print('ID is: ${item.id}');
    print('Begin is: ${item.range.begin} and End is: ${item.range.end}');
    for (String line in item.lines) {
      print(line);
    }
    print('----');
  }
}

```


CHANGELOG.md
## 2.0.1
* Downgrade meta from 1.9.0 to 1.8.0 for Flutter.

## 2.0.0
* Migrate to null safety.

## 1.1.0+5
* maintenance

## 1.1.0+4
* buggy revision reverted - no wrapperDetail

## 1.1.0+3
* i/u/b init. false, htmlCodeWrapper detailed

## 1.1.0+2
* range.duration's type to Duration

## 1.1.0+1
* maintenance +@visibleForTesting

## 1.1.0
* parses html & coordination

## 1.0.1+1
* maintenance 

## 1.0.1
* updated the package information. 

## 1.0.0
* Initial release. 



test/srt_parser_test.dart
import 'dart:async';
import 'package:test/test.dart';
import '../lib/srt_parser_2.dart';

void main() {
  const String data = '''1
00:02:26,407 --> 00:02:31,356  X1:100 X2:500 Y1:100 Y2:100
time to move on<u><b><font color="#00ff00">1@2.com</font></b></u><b><u>kriminella</u></b> aksdjf <font color="red">1@2.com</font>dfsfs<u>kriminella</u><font color="rgb(255,0,0)">1@2.com</font><font color="rgba(255,0,0,0.1)">mean-machine</font><b><i><u>kriminella beteende och foersvinnade.</u></i></b>
<b><font color="#4c3051">Detta handlar om min storebrors</font></b>

2
00:02:31,567 --> 00:02:37,164 
Vi talar inte laengre om Wade. Det aer 
som om han aldrig hade existerat.
''';
  test(
    'testing line splitter',
    () {
      final List<String> lines = splitIntoLines(data);
      expect(lines.length, 9);
      expect(lines.first, '1');
      expect(
        lines.last,
        allOf(
          [
            'som om han aldrig hade existerat.',
            isNotNull,
            const TypeMatcher<String>()
          ],
        ),
      );
    },
  );
  test(
    'chunks',
    () {
      final List<List<String>> chunks =
          splitByEmptyLine(['a', 'b', 'c', '', '123', 'adfasfdsa']);
      expect(chunks.length, 2);
      expect(chunks[0][2], 'c');
      expect(chunks[1][0], '123');
    },
  );
  test(
    'parseBeginEnd',
    () {
      final Range? result = parseBeginEnd('00:02:31,567 --> 00:02:37,164');
      expect(result?.begin, 151567);
      expect(result?.end, 157164);
    },
  );

  test(
    'out of range timeStamp',
    () {
      expect(Future<void>(
        () {
          timeStampToMillis(24, 59, 59, 999);
        },
      ), throwsA(isRangeError));
    },
  );

  test(
    'out of range minus',
    () {
      expect(Future<void>(
        () {
          timeStampToMillis(-1, 59, 59, 999);
        },
      ), throwsA(isRangeError));
    },
  );

  test(
    'parseSrt',
    () {
      final List<Subtitle> parsedSubtitle = parseSrt(data);
      expect(parsedSubtitle.length, 2);
      expect(parsedSubtitle[1].range.begin, 151567);
      expect(
          parsedSubtitle[0]
              .parsedLines[0]
              .subLines[1]
              .htmlCode
              .fontColor
              ?.argbValue,
          65280);
      expect(
          parsedSubtitle[0].parsedLines[0].subLines[2].rawString, 'kriminella');
      expect(parsedSubtitle[0].parsedLines[0].subLines[0].htmlCode.b, null);
      expect(parsedSubtitle[0].parsedLines[1].coordinates?.x, 500);
      expect(parsedSubtitle[0].parsedLines[0].subLines[2].htmlCode.b, true);
      expect(parsedSubtitle[0].parsedLines[0].subLines[2].htmlCode.u, true);
      expect(parsedSubtitle[0].parsedLines[0].subLines[2].htmlCode.i, null);
    },
  );
}


lib/color_map.dart
import 'package:csslib/parser.dart';

Map<String, Color> colorMap = {
  'transparent': const Color.hex("00ffffff"),
  'aliceBlue': const Color.hex("0f08ff"),
  'antiqueWhite': const Color.hex("0faebd7"),
  'aqua': const Color.hex("00ffff"),
  'aquaMarine': const Color.hex("7fffd4"),
  'azure': const Color.hex("f0ffff"),
  'beige': const Color.hex("f5f5dc"),
  'bisque': const Color.hex("ffe4c4"),
  'black': const Color.hex("000000"),
  'blanchedAlmond': const Color.hex("ffebcd"),
  'blue': const Color.hex("0000ff"),
  'blueViolet': const Color.hex("8a2be2"),
  'brown': const Color.hex("a52a2a"),
  'burlyWood': const Color.hex("deb887"),
  'cadetBlue': const Color.hex("5f9ea0"),
  'chartreuse': const Color.hex("7fff00"),
  'chocolate': const Color.hex("d2691e"),
  'coral': const Color.hex("ff7f50"),
  'cornFlowerBlue': const Color.hex("6495ed"),
  'cornSilk': const Color.hex("fff8dc"),
  'crimson': const Color.hex("dc143c"),
  'cyan': const Color.hex("00ffff"),
  'darkBlue': const Color.hex("00008b"),
  'darkCyan': const Color.hex("008b8b"),
  'darkGoldenRod': const Color.hex("b8860b"),
  'darkGray': const Color.hex("a9a9a9"),
  'darkGreen': const Color.hex("006400"),
  'darkGrey': const Color.hex("a9a9a9"),
  'darkKhaki': const Color.hex("bdb76b"),
  'darkMagenta': const Color.hex("8b008b"),
  'darkOliveGreen': const Color.hex("556b2f"),
  'darkOrange': const Color.hex("ff8c00"),
  'darkOrchid': const Color.hex("9932cc"),
  'darkRed': const Color.hex("8b0000"),
  'darkSalmon': const Color.hex("e9967a"),
  'darkSeaGreen': const Color.hex("8fbc8f"),
  'darkSlateBlue': const Color.hex("483d8b"),
  'darkSlateGray': const Color.hex("2f4f4f"),
  'darkSlateGrey': const Color.hex("2f4f4f"),
  'darkTurquoise': const Color.hex("00ced1"),
  'darkViolet': const Color.hex("9400d3"),
  'deepPink': const Color.hex("ff1493"),
  'deepSkyBlue': const Color.hex("00bfff"),
  'dimGray': const Color.hex("696969"),
  'dimGrey': const Color.hex("696969"),
  'dodgerBlue': const Color.hex("1e90ff"),
  'fireBrick': const Color.hex("b22222"),
  'floralWhite': const Color.hex("fffaf0"),
  'forestGreen': const Color.hex("228b22"),
  'fuchsia': const Color.hex("ff00ff"),
  'gainsboro': const Color.hex("dcdcdc"),
  'ghostWhite': const Color.hex("f8f8ff"),
  'gold': const Color.hex("ffd700"),
  'goldenRod': const Color.hex("daa520"),
  'gray': const Color.hex("808080"),
  'green': const Color.hex("008000"),
  'greenYellow': const Color.hex("adff2f"),
  'grey': const Color.hex("808080"),
  'honeydew': const Color.hex("f0fff0"),
  'hotPink': const Color.hex("ff69b4"),
  'indianRed': const Color.hex("cd5c5c"),
  'indigo': const Color.hex("4b0082"),
  'ivory': const Color.hex("fffff0"),
  'khaki': const Color.hex("f0e68c"),
  'lavender': const Color.hex("e6e6fa"),
  'lavenderBlush': const Color.hex("fff0f5"),
  'lawnGreen': const Color.hex("7cfc00"),
  'lemonChiffon': const Color.hex("fffacd"),
  'lightBlue': const Color.hex("add8e6"),
  'lightCoral': const Color.hex("f08080"),
  'lightCyan': const Color.hex("e0ffff"),
  'lightGoldenRodYellow': const Color.hex("fafad2"),
  'lightGray': const Color.hex("d3d3d3"),
  'lightGreen': const Color.hex("90ee90"),
  'lightGrey': const Color.hex("d3d3d3"),
  'lightPink': const Color.hex("ffb6c1"),
  'lightSalmon': const Color.hex("ffa07a"),
  'lightSeaGreen': const Color.hex("20b2aa"),
  'lightSkyBlue': const Color.hex("87cefa"),
  'lightSlateGray': const Color.hex("778899"),
  'lightSlateGrey': const Color.hex("778899"),
  'lightSteelBlue': const Color.hex("b0c4de"),
  'lightYellow': const Color.hex("ffffe0"),
  'lime': const Color.hex("00ff00"),
  'limeGreen': const Color.hex("32cd32"),
  'linen': const Color.hex("faf0e6"),
  'magenta': const Color.hex("ff00ff"),
  'maroon': const Color.hex("800000"),
  'mediumAquaMarine': const Color.hex("66cdaa"),
  'mediumBlue': const Color.hex("0000cd"),
  'mediumOrchid': const Color.hex("ba55d3"),
  'mediumPurple': const Color.hex("9370db"),
  'mediumSeaGreen': const Color.hex("3cb371"),
  'mediumSlateBlue': const Color.hex("7b68ee"),
  'mediumSpringGreen': const Color.hex("00fa9a"),
  'mediumTurquoise': const Color.hex("48d1cc"),
  'mediumVioletRed': const Color.hex("c71585"),
  'midnightBlue': const Color.hex("191970"),
  'mintCream': const Color.hex("f5fffa"),
  'mistyRose': const Color.hex("ffe4e1"),
  'moccasin': const Color.hex("ffe4b5"),
  'navajoWhite': const Color.hex("ffdead"),
  'navy': const Color.hex("000080"),
  'oldLace': const Color.hex("fdf5e6"),
  'olive': const Color.hex("808000"),
  'oliveDrab': const Color.hex("6b8e23"),
  'orange': const Color.hex("ffa500"),
  'orangeRed': const Color.hex("ff4500"),
  'orchid': const Color.hex("da70d6"),
  'paleGoldenRod': const Color.hex("eee8aa"),
  'paleGreen': const Color.hex("98fb98"),
  'paleTurquoise': const Color.hex("afeeee"),
  'paleVioletRed': const Color.hex("db7093"),
  'papayaWhip': const Color.hex("ffefd5"),
  'peachPuff': const Color.hex("ffdab9"),
  'peru': const Color.hex("cd85ef"),
  'pink': const Color.hex("ffc0cb"),
  'plum': const Color.hex("dda0dd"),
  'powderBlue': const Color.hex("b0e0e6"),
  'purple': const Color.hex("800080"),
  'red': const Color.hex("ff0000"),
  'rosyBrown': const Color.hex("bc8f8f"),
  'royalBlue': const Color.hex("4169e1"),
  'saddleBrown': const Color.hex("8b4513"),
  'salmon': const Color.hex("fa8072"),
  'sandyBrown': const Color.hex("f4a460"),
  'seaGreen': const Color.hex("2e8b57"),
  'seashell': const Color.hex("fff5ee"),
  'sienna': const Color.hex("a0522d"),
  'silver': const Color.hex("c0c0c0"),
  'skyBlue': const Color.hex("87ceeb"),
  'slateBlue': const Color.hex("6a5acd"),
  'slateGray': const Color.hex("708090"),
  'slateGrey': const Color.hex("708090"),
  'snow': const Color.hex("fffafa"),
  'springGreen': const Color.hex("00ff7f"),
  'steelBlue': const Color.hex("4682b4"),
  'tan': const Color.hex("d2b48c"),
  'teal': const Color.hex("008080"),
  'thistle': const Color.hex("d8bfd8"),
  'tomato': const Color.hex("ff6347"),
  'turquoise': const Color.hex("40e0d0"),
  'violet': const Color.hex("ee82ee"),
  'wheat': const Color.hex("f5deb3"),
  'white': const Color.hex("ffffff"),
  'whiteSmoke': const Color.hex("f5f5f5"),
  'yellow': const Color.hex("ffff00"),
  'yellowGreen': const Color.hex("9acd32")
};


lib/srt_parser.dart
import 'package:csslib/parser.dart';
import 'package:meta/meta.dart';
import 'package:srt_parser/color_map.dart';
import 'dart:convert' show LineSplitter;

/// formatting (partial compliance) : https://en.wikipedia.org/wiki/SubRip#Formatting
class Range {
  Range(this.begin, this.end);

  int begin;
  int end;

  Duration get duration => Duration(milliseconds: end - begin);
}

class HtmlCode {
  bool i = false;
  bool u = false;
  bool b = false;
  Color fontColor = Color.black;
}

class Coordinates {
  Coordinates({this.x, this.y});

  final int x;
  final int y;
}

class Subtitle {
  int id;
  Range range;
  List<Line> parsedLines = [];
  List<String> rawLines = [];
}

class Line {
  Line(this.rawLine);

  final String rawLine;
  Coordinates coordinates;

  // TODO(Arman):Either a whole line has code or subLines have or none
  List<SubLine> subLines = [];
}

class SubLine {
  SubLine({this.rawString});

  HtmlCode htmlCode = HtmlCode();
  String rawString;
}

@visibleForTesting
void parseHtml(Subtitle subtitle) {
  //https://regex101.com/r/LtkFNE/4
  final RegExp detectAll = RegExp(
      r'((<(b|i|u|(font color="((#([0-9a-fA-F]+))|((rgb|rgba)\(((\d{1,3}),(\d{1,3}),(\d{1,3})|(\d{1,3}),(\d{1,3}),(\d{1,3}),(0?\.[1-9]{1,2}|1))\))|([a-z]+))"))>)+)([^<|>|\/]+)((<\/(b|i|u|font)>)+)+|([^<|>|\/]+)');

  final RegExp detectFont = RegExp(r'(<font color=")');
  final RegExp detectI = RegExp(r'(<i>)');
  final RegExp detectB = RegExp(r'(<b>)');
  final RegExp detectU = RegExp(r'(<u>)');

  for (String line in subtitle.rawLines) {
    int index = subtitle.rawLines.indexOf(line);
    Iterable<Match> allMatches = detectAll.allMatches(line);

    for (Match match in allMatches) {
      String firstMatch = match.group(1);
      // not coded text
      if (match.group(23) != null) {
        subtitle.parsedLines[index].subLines
            .add(SubLine(rawString: match.group(23)));
        continue;
      }
      //Html-coded text
      else {
        SubLine subLineWithCode = SubLine();

        if (detectI.hasMatch(firstMatch)) {
          subLineWithCode.htmlCode.i = true;
        }

        if (detectB.hasMatch(firstMatch)) {
          subLineWithCode.htmlCode.b = true;
        }
        if (detectU.hasMatch(firstMatch)) {
          subLineWithCode.htmlCode.u = true;
        }
        //font color
        if (detectFont.hasMatch(firstMatch)) {
          //hexColor
          if (match.group(7) != null) {
            subLineWithCode.htmlCode.fontColor = Color.hex(match.group(7));
          }
          //rgb or rgba
          if (match.group(8) != null) {
            if (match.group(9) == 'rgb') {
              subLineWithCode.htmlCode.fontColor = Color.createRgba(
                  int.parse(match.group(11)),
                  int.parse(match.group(12)),
                  int.parse(match.group(13)));
            }
            if (match.group(9) == 'rgba') {
              subLineWithCode.htmlCode.fontColor = Color.createRgba(
                  int.parse(match.group(14)),
                  int.parse(match.group(15)),
                  int.parse(match.group(16)),
                  num.parse(match.group(17)));
            }
          }

          // if color word names
          if (match.group(18) != null) {
            subLineWithCode.htmlCode.fontColor = colorMap.entries
                .firstWhere((MapEntry entry) => entry.key == match.group(18))
                .value;
          }
        }
        subLineWithCode.rawString = match.group(19);

        subtitle.parsedLines[index].subLines.add(subLineWithCode);
      }
    }
  }
}

@visibleForTesting
void parseCoordinates(Subtitle subtitle, String chunk1) {
  final RegExp detectCoordination = RegExp(r'((X|Y)(\d)):(\d\d\d)');

  final Iterable<Match> result = detectCoordination.allMatches(chunk1);

  if (result.length != 0) {
    List listOfXs =
        result.where((Match match) => match.group(2) == 'X').toList();

    //divide by 2 and create a Coordination of each X:Y group
    for (Match item in listOfXs) {
      int number = int.parse(item.group(3));
      Match matchingY = result.firstWhere((Match matchY) {
        return (matchY.group(2) == 'Y' && int.parse(matchY.group(3)) == number);
      });

      Line parsedLine = Line(subtitle.rawLines[listOfXs.indexOf(item)]);
      parsedLine.coordinates = Coordinates(
          x: int.parse(item.group(4)), y: int.parse(matchingY.group(4)));

      subtitle.parsedLines.add(parsedLine);
    }
  } else {
    for (String line in subtitle.rawLines) {
      Line parsedLine = Line(line);
      subtitle.parsedLines.add(parsedLine);
    }
  }
}

@visibleForTesting
Range parseBeginEnd(String line) {
  final RegExp pattern = RegExp(
      r'(\d\d):(\d\d):(\d\d),(\d\d\d) --> (\d\d):(\d\d):(\d\d),(\d\d\d)');
  final Match match = pattern.firstMatch(line);

  if (match == null) {
    return null;
  } else if (int.parse(match.group(1)) > 23 ||
      int.parse(match.group(2)) > 59 ||
      int.parse(match.group(3)) > 59 ||
      int.parse(match.group(4)) > 999 ||
      int.parse(match.group(5)) > 23 ||
      int.parse(match.group(6)) > 59 ||
      int.parse(match.group(7)) > 59 ||
      int.parse(match.group(8)) > 999) {
    throw RangeError(
        'time components are out of range. Please modify the .srt file.');
  } else {
    final int begin = timeStampToMillis(
        int.parse(match.group(1)),
        int.parse(match.group(2)),
        int.parse(match.group(3)),
        int.parse(match.group(4)));

    final int end = timeStampToMillis(
        int.parse(match.group(5)),
        int.parse(match.group(6)),
        int.parse(match.group(7)),
        int.parse(match.group(8)));

    return Range(begin, end);
  }
}

@visibleForTesting
int timeStampToMillis(int hour, int minute, int sec, int ms) {
  if (hour <= 23 &&
      hour >= 0 &&
      minute <= 59 &&
      minute >= 0 &&
      sec <= 59 &&
      sec >= 0 &&
      ms <= 999 &&
      ms >= 0) {
    int result = ms;
    result += sec * 1000;
    result += minute * 60 * 1000;
    result += hour * 60 * 60 * 1000;
    return result;
  } else {
    throw RangeError('sth. is outa range');
  }
}

@visibleForTesting
List<String> splitIntoLines(String data) {
  return LineSplitter().convert(data);
}

//splits
@visibleForTesting
List<List<String>> splitByEmptyLine(List<String> lines) {
  final List<List<String>> result = [];
  List<String> chunk = <String>[];

  for (String line in lines) {
    if (line.isEmpty) {
      result.add(chunk);
      chunk = [];
    } else {
      chunk.add(line);
    }
  }
  if (chunk.isNotEmpty) {
    result.add(chunk);
  }

  return result;
}

List<Subtitle> parseSrt(String srt) {
  final List<Subtitle> result = [];

  final List<String> split = splitIntoLines(srt);
  final List<List<String>> splitChunk = splitByEmptyLine(split);

  for (List<String> chunk in splitChunk) {
    final Subtitle subtitle = Subtitle();
    subtitle.id = int.parse(chunk[0]);
    subtitle.range = parseBeginEnd(chunk[1]);
    subtitle.rawLines = chunk.sublist(2);
    parseCoordinates(subtitle, chunk[1]);
    parseHtml(subtitle);
    result.add(subtitle);
  }

  return result;
}


lib/srt_parser_2.dart
import 'dart:convert' show LineSplitter;

import 'package:collection/collection.dart';
import 'package:csslib/parser.dart';
import 'package:meta/meta.dart';

import './color_map.dart';

/// formatting (partial compliance) : https://en.wikipedia.org/wiki/SubRip#Formatting
class Range {
  const Range(this.begin, this.end);

  final int begin;
  final int end;

  Duration get duration => Duration(milliseconds: end - begin);
}

class HtmlCode {
  bool? i;
  bool? u;
  bool? b;
  Color? fontColor;
}

class Coordinates {
  const Coordinates({required this.x, required this.y});

  final int x;
  final int y;
}

class Subtitle {
  Subtitle({
    required this.id,
    required this.range,
    required this.rawLines,
  });

  final int id;
  final Range range;
  final List<Line> parsedLines = [];
  final List<String> rawLines;
}

class Line {
  Line(this.rawLine, {this.coordinates});

  final String rawLine;
  final Coordinates? coordinates;

  // TODO(Arman):Either a whole line has code or subLines have or none
  final List<SubLine> subLines = [];
}

class SubLine {
  SubLine({this.rawString});

  final HtmlCode htmlCode = HtmlCode();
  final String? rawString;
}

@visibleForTesting
void parseHtml(Subtitle subtitle) {
  //https://regex101.com/r/LtkFNE/4
  final RegExp detectAll = RegExp(
      r'((<(b|i|u|(font color="((#([0-9a-fA-F]+))|((rgb|rgba)\(((\d{1,3}),(\d{1,3}),(\d{1,3})|(\d{1,3}),(\d{1,3}),(\d{1,3}),(0?\.[1-9]{1,2}|1))\))|([a-z]+))"))>)+)([^<|>|\/]+)((<\/(b|i|u|font)>)+)+|([^<|>|\/]+)');

  final RegExp detectFont = RegExp(r'(<font color=")');
  final RegExp detectI = RegExp(r'(<i>)');
  final RegExp detectB = RegExp(r'(<b>)');
  final RegExp detectU = RegExp(r'(<u>)');

  for (String line in subtitle.rawLines) {
    int index = subtitle.rawLines.indexOf(line);
    Iterable<Match> allMatches = detectAll.allMatches(line);

    for (Match match in allMatches) {
      String? firstMatch = match.group(1);
      // not coded text
      if (match.group(23) != null) {
        subtitle.parsedLines[index].subLines
            .add(SubLine(rawString: match.group(23)!));
        continue;
      }
      //Html-coded text
      else if (firstMatch != null) {
        SubLine subLineWithCode = SubLine(rawString: match.group(19));

        if (detectI.hasMatch(firstMatch)) {
          subLineWithCode.htmlCode.i = true;
        }

        if (detectB.hasMatch(firstMatch)) {
          subLineWithCode.htmlCode.b = true;
        }
        if (detectU.hasMatch(firstMatch)) {
          subLineWithCode.htmlCode.u = true;
        }
        //font color
        if (detectFont.hasMatch(firstMatch)) {
          //hexColor
          if (match.group(7) != null) {
            subLineWithCode.htmlCode.fontColor = Color.hex(match.group(7)!);
          }
          //rgb or rgba
          if (match.group(8) != null) {
            if (match.group(9) == 'rgb' &&
                match.groups([11, 12, 13]).every((g) => g != null)) {
              subLineWithCode.htmlCode.fontColor = Color.createRgba(
                  int.parse(match.group(11)!),
                  int.parse(match.group(12)!),
                  int.parse(match.group(13)!));
            }
            if (match.group(9) == 'rgba' &&
                match.groups([14, 15, 16, 17]).every((g) => g != null)) {
              subLineWithCode.htmlCode.fontColor = Color.createRgba(
                  int.parse(match.group(14)!),
                  int.parse(match.group(15)!),
                  int.parse(match.group(16)!),
                  num.parse(match.group(17)!));
            }
          }

          // if color word names
          if (match.group(18) != null) {
            subLineWithCode.htmlCode.fontColor = colorMap.entries
                .firstWhere((MapEntry entry) => entry.key == match.group(18))
                .value;
          }
        }

        subtitle.parsedLines[index].subLines.add(subLineWithCode);
      }
    }
  }
}

@visibleForTesting
void parseCoordinates(Subtitle subtitle, String chunk1) {
  final RegExp detectCoordination = RegExp(r'((X|Y)(\d)):(\d\d\d)');

  final Iterable<Match> result = detectCoordination.allMatches(chunk1);

  if (result.length != 0) {
    List<Match> listOfXs =
        result.where((Match match) => match.group(2) == 'X').toList();

    //divide by 2 and create a Coordination of each X:Y group
    for (Match item in listOfXs) {
      if (item.groups([3, 4]).any((g) => g == null)) continue;

      int number = int.parse(item.group(3)!);
      Match? matchingY = result.firstWhereOrNull((Match matchY) {
        if (matchY.group(3) == null) return false;

        return (matchY.group(2) == 'Y' &&
            int.tryParse(matchY.group(3)!) == number);
      });
      if (matchingY == null) continue;

      final coordinates = Coordinates(
          x: int.parse(item.group(4) ?? '0'),
          y: int.parse(matchingY.group(4) ?? '0'));
      Line parsedLine = Line(subtitle.rawLines[listOfXs.indexOf(item)],
          coordinates: coordinates);
      subtitle.parsedLines.add(parsedLine);
    }
  } else {
    for (String line in subtitle.rawLines) {
      Line parsedLine = Line(line);
      subtitle.parsedLines.add(parsedLine);
    }
  }
}

@visibleForTesting
Range? parseBeginEnd(String line) {
  final RegExp pattern = RegExp(
      r'(\d\d):(\d\d):(\d\d),(\d\d\d) --> (\d\d):(\d\d):(\d\d),(\d\d\d)');
  final Match? match = pattern.firstMatch(line);

  if (match == null) {
    return null;
  } else if (match.groups([1, 2, 3, 4, 5, 6, 7, 8]).any((g) => g == null) ||
      int.parse(match.group(1)!) > 23 ||
      int.parse(match.group(2)!) > 59 ||
      int.parse(match.group(3)!) > 59 ||
      int.parse(match.group(4)!) > 999 ||
      int.parse(match.group(5)!) > 23 ||
      int.parse(match.group(6)!) > 59 ||
      int.parse(match.group(7)!) > 59 ||
      int.parse(match.group(8)!) > 999) {
    throw RangeError(
        'time components are out of range. Please modify the .srt file.');
  } else {
    final int begin = timeStampToMillis(
        int.parse(match.group(1)!),
        int.parse(match.group(2)!),
        int.parse(match.group(3)!),
        int.parse(match.group(4)!));

    final int end = timeStampToMillis(
        int.parse(match.group(5)!),
        int.parse(match.group(6)!),
        int.parse(match.group(7)!),
        int.parse(match.group(8)!));

    return Range(begin, end);
  }
}

@visibleForTesting
int timeStampToMillis(int hour, int minute, int sec, int ms) {
  if (hour <= 23 &&
      hour >= 0 &&
      minute <= 59 &&
      minute >= 0 &&
      sec <= 59 &&
      sec >= 0 &&
      ms <= 999 &&
      ms >= 0) {
    int result = ms;
    result += sec * 1000;
    result += minute * 60 * 1000;
    result += hour * 60 * 60 * 1000;
    return result;
  } else {
    throw RangeError('sth. is outa range');
  }
}

@visibleForTesting
List<String> splitIntoLines(String data) {
  return LineSplitter().convert(data);
}

//splits
@visibleForTesting
List<List<String>> splitByEmptyLine(List<String> lines) {
  final List<List<String>> result = [];
  List<String> chunk = <String>[];

  for (String line in lines) {
    if (line.isEmpty) {
      result.add(chunk);
      chunk = [];
    } else {
      chunk.add(line);
    }
  }
  if (chunk.isNotEmpty) {
    result.add(chunk);
  }

  return result;
}

List<Subtitle> parseSrt(String srt) {
  final List<Subtitle> result = [];

  final List<String> split = splitIntoLines(srt);
  final List<List<String>> splitChunk = splitByEmptyLine(split);

  for (List<String> chunk in splitChunk) {
    final range = parseBeginEnd(chunk[1]);
    if (range == null) continue;
    final Subtitle subtitle = Subtitle(
      id: int.parse(chunk[0]),
      range: range,
      rawLines: chunk.sublist(2),
    );
    parseCoordinates(subtitle, chunk[1]);
    parseHtml(subtitle);
    result.add(subtitle);
  }

  return result;
}


example/example.dart
import 'package:srt_parser_2/srt_parser_2.dart';

const String data = '''1
00:02:26,407 --> 00:02:31,356  X1:100 X2:100 Y1:100 Y2:100
+ time to move on, <u><b><font color="#00ff00">Arman</font></b></u>.
- OK, will do.

2
00:02:31,567 --> 00:02:37,164 
+ Lukas is publishing his library.
- I like the man.
''';

void main() {
  List<Subtitle> subtitles = parseSrt(data);
  for (Subtitle item in subtitles) {
    print('subtitle\'s ID is: ${item.id}');
    print(
        'subtitle\'s Begin is: ${item.range.begin} and End is: ${item.range.end}');
    item.parsedLines.forEach((Line line) {
      return line.subLines.forEach((SubLine subLine) => print(
          'line${item.parsedLines.indexOf(line)} subline${line.subLines.indexOf(subLine)} is: ${subLine.rawString}'));
    });
    print('----');
  }
  if (subtitles[0].parsedLines[0].subLines[1].htmlCode.b == true) {
    print('true');
  }
}
